changefeedccl: match nil metrics behaviour

wenyihu6 opened the pull request:
Previously, `validateExternalConnectionSinkURI` would validate the changefeed
sink URI by creating a fake sink and passing `nil` for `metricsRecorder`. This
is usually not problematic since the sink is never used. A new patch, #117693,
is now changing this and calling `metricsBuilder` interface method inside
`makeKafkaSink`. To resolve this, this patch changes getSink in
`validateExternalConnectionSinkURI` to pass in `(*sliMetrics)(nil)` to avoid
calling methods on a nil interface.

See also: https://github.com/cockroachdb/cockroach/pull/117693 
Release note: none
Epic: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/118421)
<!-- Reviewable:end -->

wenyihu6 committed:
changefeedccl: match nil metrics behaviour

Previously, `validateExternalConnectionSinkURI` would validate the changefeed
sink URI by creating a fake sink and passing `nil` for `metricsRecorder`. This
is usually not problematic since the sink is never used. A new patch, #117693,
is now changing this and calling `metricsBuilder` interface method inside
`makeKafkaSink`. To resolve this, this patch changes getSink in
`validateExternalConnectionSinkURI` to pass in `(*sliMetrics)(nil)` to avoid
calling methods on a nil interface.

See also: https://github.com/cockroachdb/cockroach/pull/117693
Release note: none
Epic: none

wenyihu6 commented:
Alternatively, we can check if m is nil here and return nilMetricsRecorderBuilder(recordingRequired) if it is. wdyt?

https://github.com/cockroachdb/cockroach/blob/cb674c239069480b57ebd9ba574eb3e3907e1342/pkg/ccl/changefeedccl/sink.go#L211-L216

wenyihu6 commented:
TFTR!!

bors r=miretskiy
